{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4fac5ae-2598-4f24-92e5-5cae8604556f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "import torch\n",
    "import pandas as pd\n",
    "\n",
    "from procyon.data.inference_utils import (\n",
    "    create_caption_input_simple,\n",
    "    create_qa_input_simple,\n",
    "    uniprot_id_to_index,\n",
    "    ProCyonQAInference,\n",
    ")\n",
    "from procyon.model.model_unified import UnifiedProCyon"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64f3f1b7",
   "metadata": {},
   "source": [
    "This notebook contains a minimal example of using a pre-trained ProCyon model for phenotype generation with QA filtering, as shown in Figure 4A of the main manuscript."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49abf566",
   "metadata": {},
   "source": [
    "# Load pre-trained model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a716bbd3-d4fa-43d8-9647-eff3820c03c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace with the path where you downloaded a pre-trained ProCyon model (e.g. ProCyon-Full)\n",
    "checkpoint_path = \"~/storage/Lab/PLM/huggingface_repos/ProCyon-Full\"\n",
    "CKPT_NAME = os.path.expanduser(checkpoint_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cc758a42",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_args = torch.load(os.path.join(CKPT_NAME, \"data_args.pt\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a077a38c-cf33-48cb-93a3-e0064fd7a0f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')\n",
    "model, _ = UnifiedProCyon.from_pretrained(checkpoint_dir=CKPT_NAME)\n",
    "model.to(device)\n",
    "model.bfloat16() # Quantize the model to a smaller precision\n",
    "model.eval()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0af0ba5d",
   "metadata": {},
   "source": [
    "# Generate model input"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b844208f",
   "metadata": {},
   "source": [
    "In this example, we'll generate phenotypes for the protein AKNAD1, as shown in Figure 4A in the main text. First, we generate the input to the ProCyon model.\n",
    "\n",
    "The inputs to a ProCyon model are structed as a dict. At a high-level, these dicts support large training batches while minimizing\n",
    "data duplication within the dict. To support this, the dicts have the following structure:\n",
    "- `data` - The unique elements of the ProCyon-Instruct used in this data batch\n",
    "  - `seq` - Integer IDs of the associated proteins, as generated by `uniprot_id_to_index()`\n",
    "  - `text` - The text descriptions of phenotypes in this batch\n",
    "- `input` - The elements within `data` used in each individual input in this batch.\n",
    "  - `seq` - List of lists, where the `i`'th list is the proteins used in the `i`'th input in the batch.\n",
    "  - `text` - List of lists, where the `i`'th list is the phenotypes used in the `i`'th input in the batch.\n",
    "- `target` - Used in training only. The target output for each input in the batch.\n",
    "- `instructions` - The natural language prompt for each input, which will be populated with the appropriate texts and protein embeddings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "865a5ffb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Internally, ProCyon uses integer IDs that have been assigned to UniProt proteins in ProCyon-Instruct.\n",
    "want_proteins = [\"Q5T1N1\"]\n",
    "protein_ids = [uniprot_id_to_index(x) for x in want_proteins]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c8b35199-cac7-4485-93c7-84e6b0f4b468",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_simple = create_caption_input_simple(\n",
    "    input_aaseq_ids=protein_ids,\n",
    "    data_args=data_args,\n",
    "    # The `instruction_source_dataset` and `instruction_source_relation` here control the style\n",
    "    # of pre-templated instruction used in these queries. In particular, here we query for UniProt-style\n",
    "    # functional descriptions.\n",
    "    instruction_source_dataset=\"uniprot\",\n",
    "    instruction_source_relation=\"all\",\n",
    "    aaseq_type=\"protein\",\n",
    "    task_type=\"caption\",\n",
    "    icl_example_number=1,\n",
    "    device=device,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7bfbd57b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'data': {'seq': tensor([  5, 530], device='cuda:0'),\n",
       "  'seq_idx': tensor([  5, 530], device='cuda:0'),\n",
       "  'text': ['Adapter protein implicated in the regulation of a large spectrum of both general and specialized signaling pathways. Binds to a large number of partners, usually by recognition of a phosphoserine or phosphothreonine motif. Binding generally results in the modulation of the activity of the binding partner. Negatively regulates the kinase activity of PDPK1.'],\n",
       "  'drug': None},\n",
       " 'input': {'seq': [[0, 1]], 'text': [[0]], 'drug': None},\n",
       " 'target': {'seq': None, 'text': None, 'drug': None},\n",
       " 'instructions': [\"Definition: You will be shown a protein. Your job is to describe function(s) of a protein, and this protein should be involved in the function. This response provides any useful information about the protein, mostly biological knowledge. When a protein is a precursor of different chains/peptides having different functions, specify the role of each peptide in separate paragraphs. The same holds true for isoform-specific functions. Multifunctional enzymes should have only one 'Function' section for the different activities. When a protein displays some very different or contradictory functions depending on the context, describe them in different sentences. This case is rare.\\nPositive example 1:\\nProtein: <|protein|>\\nOutput: [ANSWER] [EXT]\\nNow, complete the following instance:\\nProtein: <|protein|>\\nOutput: [ANSWER] \"]}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_simple"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "682788ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Definition: You will be shown a protein. Your job is to describe function(s) of a protein, and this protein should be involved in the function. This response provides any useful information about the protein, mostly biological knowledge. When a protein is a precursor of different chains/peptides having different functions, specify the role of each peptide in separate paragraphs. The same holds true for isoform-specific functions. Multifunctional enzymes should have only one 'Function' section for the different activities. When a protein displays some very different or contradictory functions depending on the context, describe them in different sentences. This case is rare.\n",
      "Positive example 1:\n",
      "Protein: <|protein|>\n",
      "Output: [ANSWER] [EXT]\n",
      "Now, complete the following instance:\n",
      "Protein: <|protein|>\n",
      "Output: [ANSWER] \n"
     ]
    }
   ],
   "source": [
    "# Note the special `<|protein|>` tokens, which will be replaced with the corresponding protein's embedding.\n",
    "print(input_simple[\"instructions\"][0])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25b72d38",
   "metadata": {},
   "source": [
    "# Generate phenotypes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8fd504a",
   "metadata": {},
   "source": [
    "Now we'll use the input to generate phenotypes using the pre-trained model. We'll use diverse beam search (DBS) as the text generation method, as this allows generating multiple, distinct phenotypes for a single input. Note that we also support greedy decoding, temperature sampling, and nucleus sampling, but find that diverse beam search produces the highest quality outputs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "70b8fcd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "text_gen_args = {\n",
    "    \"method\": \"beam\",\n",
    "    # Maximum length of generated text.\n",
    "    \"max_len\": 200,\n",
    "    # Total number of beams maintained per input. `beam_size` / `beam_group_size` = number of phenotypes returned per input.\n",
    "    \"beam_size\": 20,\n",
    "    # Size of the individual beam groups in DBS.\n",
    "    \"beam_group_size\": 2,\n",
    "    # Penalty applied to repetition within a beam group.\n",
    "    \"diversity_penalty\": 0.8,\n",
    "}\n",
    "\n",
    "out_tokens, log_probs, output_logits, out_text = model.generate(\n",
    "    inputs=input_simple,\n",
    "    aaseq_type=\"protein\",\n",
    "    **text_gen_args\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "67134e66",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['Subunit of the V1 complex of vacuolar(H+)-ATPase (V-ATPase) which is required for the transport of proteins into the endoplasmic reticulum (ER) to the outer membrane.',\n",
       "  'Subunit of the V1 complex of vacuolar(H+)-ATPase (V-ATPase) which is required for the transport of proteins into the endoplasmic reticulum (ER) to the outer membrane.',\n",
       "  'May be involved in transcriptional regulation',\n",
       "  'May be involved in transcriptional regulation',\n",
       "  'Component of the cytosolic machinery in which it is required for the formation of the actin cytoskeleton. May play a role in dendrite formation. May play a role in dendrite formation.',\n",
       "  'Component of the cytosolic machinery in which it is required for the formation of the actin cytoskeleton. May play a role in dendrite formation. May play a role in dendrite formation.',\n",
       "  'Plays a role in spermatogenesis. May play a role in spermatogenesis.',\n",
       "  'Plays a role in spermatogenesis. May play a role in spermatogenesis.',\n",
       "  'Putative regulator of cAMP-dependent protein kinase activity.',\n",
       "  'Putative regulator of cAMP-dependent protein kinase activity.',\n",
       "  'Functions as an adapter protein that is implicated in the regulation of cell cycle progression.',\n",
       "  'Functions as an adapter protein that is implicated in the regulation of cell cycle progression.',\n",
       "  'Transcription factor that binds to and repress transcription of target genes.',\n",
       "  'Transcription factor that binds to and repress transcription of target genes.',\n",
       "  'Guanine nucleotide exchange factor (GEF) for RAB9A, RAB9B and RAB9B, RAB9B, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9B, RAB9B, RAB9B, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, R',\n",
       "  'Guanine nucleotide exchange factor (GEF) for RAB9A, RAB9B and RAB9B, RAB9B, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9B, RAB9B, RAB9B, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9 and R',\n",
       "  'Regulates RAC1 activity by binding to RAC1 and RAC1. Activates RAC1-mediated activation of RAC1, RAC1, RAC1, RAC1, RAC1, RAC1, RAB9, RAB9, RAB9/13, RAB9, RAB9/13, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9 and RAB9B, RAB9A and RAB9B. Inhibits RAC1-mediated activation of RAC1 and RAB9-mediated activation of RAC1. Inhibits RAC1B-mediated activation of RAC1 and RAB9-mediated activation. Inhibits RAC1-mediated activation of RAC1B/PTK2-mediated activation of RAC1. In',\n",
       "  'Regulates RAC1 activity by binding to RAC1 and RAC1. Activates RAC1-mediated activation of RAC1, RAC1, RAC1, RAC1, RAC1, RAC1, RAB9, RAB9, RAB9/13, RAB9, RAB9/13, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9 and RAB9B, RAB9A and RAB9B. Inhibits RAC1-mediated activation of RAC1 and RAB9-mediated activation of RAC1. Inhibits RAC1B-mediated activation of RAC1 and RAB9-mediated activation. Inhibits RAC1-mediated activation of RAC1B/PTK2-mediated activation of RAC1.',\n",
       "  'Inhibits cell proliferation and cell migration. Activates transcription of CDK2 and CDK2-mediated cell cycle progression by promoting CDK2-mediated cell cycle progression.',\n",
       "  'Inhibits cell proliferation and cell migration. Activates transcription of CDK2 and CDK2-mediated cell cycle progression by promoting CDK2-mediated cell cycle progression.']]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "294a85bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Subunit of the V1 complex of vacuolar(H+)-ATPase (V-ATPase) which is required for the transport of proteins into the endoplasmic reticulum (ER) to the outer membrane.',\n",
       " 'May be involved in transcriptional regulation',\n",
       " 'Component of the cytosolic machinery in which it is required for the formation of the actin cytoskeleton. May play a role in dendrite formation. May play a role in dendrite formation.',\n",
       " 'Plays a role in spermatogenesis. May play a role in spermatogenesis.',\n",
       " 'Putative regulator of cAMP-dependent protein kinase activity.',\n",
       " 'Functions as an adapter protein that is implicated in the regulation of cell cycle progression.',\n",
       " 'Transcription factor that binds to and repress transcription of target genes.',\n",
       " 'Guanine nucleotide exchange factor (GEF) for RAB9A, RAB9B and RAB9B, RAB9B, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9B, RAB9B, RAB9B, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, R',\n",
       " 'Regulates RAC1 activity by binding to RAC1 and RAC1. Activates RAC1-mediated activation of RAC1, RAC1, RAC1, RAC1, RAC1, RAC1, RAB9, RAB9, RAB9/13, RAB9, RAB9/13, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9 and RAB9B, RAB9A and RAB9B. Inhibits RAC1-mediated activation of RAC1 and RAB9-mediated activation of RAC1. Inhibits RAC1B-mediated activation of RAC1 and RAB9-mediated activation. Inhibits RAC1-mediated activation of RAC1B/PTK2-mediated activation of RAC1. In',\n",
       " 'Inhibits cell proliferation and cell migration. Activates transcription of CDK2 and CDK2-mediated cell cycle progression by promoting CDK2-mediated cell cycle progression.']"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Typically we see a lot of repetition within a DBS group, so for the final set of generated phenotypes, we take\n",
    "# the first output per group.\n",
    "output_phenotypes = [\n",
    "    phen for i, phen in enumerate(out_text[0]) if i % text_gen_args[\"beam_group_size\"] == 0\n",
    "]\n",
    "output_phenotypes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1747379a",
   "metadata": {},
   "source": [
    "# QA filtering of generated phenotypes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b701deb",
   "metadata": {},
   "source": [
    "Next, we'll leverage ProCyon's question-answering capabilities to further filter down the generated phenotypes to a set of high-confidence outputs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "afdb2edf-7d7b-40c3-b5ee-2afeb3353cc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Wrap the pre-trained model with a helper class specifically for QA.\n",
    "qa_model = ProCyonQAInference(model, device=device)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f0b1d04b-9390-4e6d-86aa-fd9428118626",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TEXT 0 --------------------------------------------\n",
      "Subunit of the V1 complex of vacuolar(H+)-ATPase (V-ATPase) which is required for the transport of proteins into the endoplasmic reticulum (ER) to the outer membrane.\n",
      "Yes: 0.173\n",
      "No: 0.827\n",
      "TEXT 1 --------------------------------------------\n",
      "May be involved in transcriptional regulation\n",
      "Yes: 0.015\n",
      "No: 0.985\n",
      "TEXT 2 --------------------------------------------\n",
      "Component of the cytosolic machinery in which it is required for the formation of the actin cytoskeleton. May play a role in dendrite formation. May play a role in dendrite formation.\n",
      "Yes: 0.860\n",
      "No: 0.140\n",
      "TEXT 3 --------------------------------------------\n",
      "Plays a role in spermatogenesis. May play a role in spermatogenesis.\n",
      "Yes: 0.867\n",
      "No: 0.133\n",
      "TEXT 4 --------------------------------------------\n",
      "Putative regulator of cAMP-dependent protein kinase activity.\n",
      "Yes: 0.827\n",
      "No: 0.173\n",
      "TEXT 5 --------------------------------------------\n",
      "Functions as an adapter protein that is implicated in the regulation of cell cycle progression.\n",
      "Yes: 0.679\n",
      "No: 0.321\n",
      "TEXT 6 --------------------------------------------\n",
      "Transcription factor that binds to and repress transcription of target genes.\n",
      "Yes: 0.234\n",
      "No: 0.766\n",
      "TEXT 7 --------------------------------------------\n",
      "Guanine nucleotide exchange factor (GEF) for RAB9A, RAB9B and RAB9B, RAB9B, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9B, RAB9B, RAB9B, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, R\n",
      "Yes: 0.500\n",
      "No: 0.500\n",
      "TEXT 8 --------------------------------------------\n",
      "Regulates RAC1 activity by binding to RAC1 and RAC1. Activates RAC1-mediated activation of RAC1, RAC1, RAC1, RAC1, RAC1, RAC1, RAB9, RAB9, RAB9/13, RAB9, RAB9/13, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9, RAB9 and RAB9B, RAB9A and RAB9B. Inhibits RAC1-mediated activation of RAC1 and RAB9-mediated activation of RAC1. Inhibits RAC1B-mediated activation of RAC1 and RAB9-mediated activation. Inhibits RAC1-mediated activation of RAC1B/PTK2-mediated activation of RAC1. In\n",
      "Yes: 0.706\n",
      "No: 0.294\n",
      "TEXT 9 --------------------------------------------\n",
      "Inhibits cell proliferation and cell migration. Activates transcription of CDK2 and CDK2-mediated cell cycle progression by promoting CDK2-mediated cell cycle progression.\n",
      "Yes: 0.755\n",
      "No: 0.245\n"
     ]
    }
   ],
   "source": [
    "# Try QA filtering\n",
    "results = []\n",
    "for i, query_text in enumerate(output_phenotypes):\n",
    "    input_qa_simple = create_qa_input_simple(\n",
    "        input_aaseq_ids=protein_ids,\n",
    "        data_args=data_args,\n",
    "        input_description=query_text,\n",
    "        instruction_source_dataset=\"uniprot\",\n",
    "        instruction_source_relation=\"all\",\n",
    "        aaseq_type=\"protein\",\n",
    "        icl_example_number=1,\n",
    "        device=device,\n",
    "    )\n",
    "\n",
    "    with torch.no_grad():\n",
    "        model_qa_out = qa_model(input_qa_simple)\n",
    "\n",
    "    yes_prob = model_qa_out[\"pred\"][0, qa_model.yes_token].item()\n",
    "    no_prob = model_qa_out[\"pred\"][0, qa_model.no_token].item()\n",
    "\n",
    "\n",
    "    print(f\"TEXT {i} --------------------------------------------\")\n",
    "    print(query_text)\n",
    "    print(f\"Yes: {yes_prob:0.3f}\")\n",
    "    print(f\"No: {no_prob:0.3f}\")\n",
    "\n",
    "    results.append({\n",
    "        \"phenotype\": query_text,\n",
    "        \"yes_prob\": yes_prob\n",
    "    })\n",
    "\n",
    "results = pd.DataFrame(results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c8cd7bbf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype</th>\n",
       "      <th>yes_prob</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Component of the cytosolic machinery in which ...</td>\n",
       "      <td>0.859625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Plays a role in spermatogenesis. May play a ro...</td>\n",
       "      <td>0.866983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Putative regulator of cAMP-dependent protein k...</td>\n",
       "      <td>0.826651</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           phenotype  yes_prob\n",
       "2  Component of the cytosolic machinery in which ...  0.859625\n",
       "3  Plays a role in spermatogenesis. May play a ro...  0.866983\n",
       "4  Putative regulator of cAMP-dependent protein k...  0.826651"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results.query(\"yes_prob >= 0.8\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
